### PR TITLE
Autocomplete: migrate to `bfg/contextForIdentifiers`

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

- Added `getGraphContextIdentifiers` to the tree-sitter SDK, which will be used to experiment with graph context.
    - For now, only JS and TS queries are polished. Going to improve `go` and `python` in a separate PR after experimenting with JS.
- Migrated to the new `bfg/contextForIdentifiers` BFG request, where clients are responsible for specifying the list of identifiers. More context https://github.com/sourcegraph/bfg-private/pull/171
    - The BFG request migration won't be an immediate priority. [We plan to experiment with LSP context in the short term](https://docs.google.com/document/d/1V2SB-ixi8OrhTwCMPO0MseBkA4tnipAkse5CCXbucXI/edit?usp=sharing). I want to merge graph-context-identifiers tree-sitter queries and start using them in combination with LSP. 
- Updated the `execQueryWrapper` API to allow different argument combinations.
- Added the tree-sitter tree representation output to snapshot tests if `// only` is used to debug a particular test.

## Test plan

- CI
- Added graph-identifiers query snapshot tests

### CODE REVIEW

**Comment by valerybugakov** [comment]
Tagging @philipp-spiess for early 👀

File: vscode/src/completions/context/context-mixer.ts [hunk 1]
```diff
@@ -61,6 +64,10 @@ export interface GetContextResult {
 export class ContextMixer implements vscode.Disposable {
     constructor(private strategyFactory: ContextStrategyFactory) {}
 
+    public isBfgRetrieverEnabled(): boolean {
+        return this.strategyFactory.graphRetriever instanceof BfgRetriever
+    }
```
Comment (philipp-spiess) [inline vscode/src/completions/context/context-mixer.ts:69]: Unused?

File: vscode/src/completions/context/context-strategy.ts [hunk 2]
```diff
@@ -16,13 +16,14 @@ export type ContextStrategy =
...
     getStrategy(document: vscode.TextDocument): { name: ContextStrategy; retrievers: ContextRetriever[] }
+    graphRetriever: ContextRetriever | undefined
 }
 
 export class DefaultContextStrategyFactory implements ContextStrategyFactory {
     private disposables: vscode.Disposable[] = []
 
     private localRetriever: ContextRetriever | undefined
-    private graphRetriever: ContextRetriever | undefined
+    public graphRetriever: ContextRetriever | undefined
```
Comment (philipp-spiess) [inline vscode/src/completions/context/context-strategy.ts:26]: Why is this made public? 🤔

File: vscode/src/completions/context/retrievers/bfg/bfg-retriever.test.ts
Comment (philipp-spiess) [inline vscode/src/completions/context/retrievers/bfg/bfg-retriever.test.ts:1]: This is just a copy for now, or? I don't see know if there's changed behavior here.

File: vscode/src/completions/context/retrievers/bfg/bfg-retriever.ts [hunk 3]
```diff
@@ -184,32 +188,57 @@ export class BfgRetriever implements ContextRetriever {
...
+                new Set([
+                    // Get last 10 identifiers from the last candidate insert text
+                    // in hopes that LLM partially guessed the right completion.
+                    ...getLastNBfgIdentifiersFromString({
+                        n: 10,
+                        document,
+                        position,
+                        currentLinePrefix: docContext.currentLinePrefix,
+                        source: lastCandidateCurrentLine,
+                    }),
```
Comment (philipp-spiess) [inline vscode/src/completions/context/retrievers/bfg/bfg-retriever.ts:206]: Hm interesting so this means BFG will only be good if we have two round trips to the LLM? One to get a good _partial suggestion_ and then one that will include BFG?

File: vscode/src/completions/doc-context-getters.ts [hunk 4]
```diff
@@ -52,7 +55,83 @@ export function getCompletionIntent(params: GetCompletionIntentParams): Completi
...
+    position: vscode.Position
+    currentLinePrefix: string
+    /**
+     * Parse this source string to get the tree for the tree-sitter query
+     * instead of using `document.getText`
+     */
+    source?: string
+}
+
+export function getLastNBfgIdentifiersFromDocument(
```
Comment (philipp-spiess) [inline vscode/src/completions/doc-context-getters.ts:76]: Should we couple this to BFG or just rename it to `getLastNIdentifier` instead?

If this is BFG specific maybe we should put it closer to the BFG parts in the source code?

File: vscode/src/completions/doc-context-getters.ts [hunk 5]
```diff
@@ -52,7 +55,83 @@ export function getCompletionIntent(params: GetCompletionIntentParams): Completi
...
+
+interface GetLastNBfgIdentifiersFromStringParams {
+    n: number
+    document: vscode.TextDocument
+    position: vscode.Position
+    currentLinePrefix: string /**
+     * Parse this source string to get the tree for the tree-sitter query
+     * instead of using `document.getText()`
+     */
+    source: string
```
Comment (philipp-spiess) [inline vscode/src/completions/doc-context-getters.ts:107]: docstring indentation gone haywire :D

File: vscode/src/completions/get-inline-completions.ts [hunk 6]
```diff
@@ -262,7 +262,12 @@ async function doGetInlineCompletions(
                   handleDidPartiallyAcceptCompletionItem,
               })
             : null
-    if (resultToReuse) {
+
+    // If smart throttle is enabled we want to reuse the last candidate only if the request is throttled (see below).
+    // Otherwise we want to make a new request based on the latest user input and update the completion.
+    // We also use the last candidate insert text as a source for identifiers for the updated context to improve the
+    // updated completion.
```
Comment (philipp-spiess) [inline vscode/src/completions/get-inline-completions.ts:269]: I’m not sure I understand the intent behind this. Looking at the smart throttle mess I produced it seems that it only returns `null` if the completion was aborted within the smart throttle interval.

It also means that we only return the `resultToReuse` after a 25ms sleep time and no longer instantly.

Maybe that's expected; If so: do you mind elaborating?

**Review by philipp-spiess — COMMENTED** [review]
Good stuff! Left some thoughts inline

File: vscode/src/completions/context/retrievers/bfg/bfg-retriever.test.ts
Comment (valerybugakov) [reply vscode/src/completions/context/retrievers/bfg/bfg-retriever.test.ts:1]: Yep. I'll delete it for now and return to it when we shift the focus back to BFG.

File: vscode/src/completions/context/retrievers/bfg/bfg-retriever.ts (hunk 3 above)
Comment (valerybugakov) [reply vscode/src/completions/context/retrievers/bfg/bfg-retriever.ts:206]: The first request will use identifiers from the last N lines. The second one will enhanced by the identifiers extracted from the last candidate.

File: vscode/src/completions/doc-context-getters.ts (hunk 4 above)
Comment (valerybugakov) [reply vscode/src/completions/doc-context-getters.ts:80]: Yes! Moved it closer to context retrievers

**Comment by valerybugakov** [comment]
Please take a look, folks! @umpox, @philipp-spiess

**Review by philipp-spiess — APPROVED** [review]
Noice

**Pull request merged**